wake_t/physics_models/laser/envelope_solver.py

```diff
@@ -0,0 +1,281 @@
+        d_theta1 -= 2 * np.pi
+    if d_theta2 > 1.5 * np.pi:
+        d_theta2 -= 2 * np.pi
+
+    return 1.5 * d_theta1 / dz - 0.5 * d_theta2 / dz
+
+
+@njit()
+def chi():
+    return 0
+
+
+@njit()
+def rhs(a_old, a, a_new, j, dz, k, dr, nr, dt, k0p, th, th1, th2):
+    """
+    The right-hand side of equation 7 in Benedetti, 2018.
+
+    Parameters
+    ----------
+    a_old : array
+        The array of the values of â at former time step. Dimension: (nz+2)*nr
+    a : array
+        The array of the values of â at current time step. Dimension: (nz+2)*nr
+    a_new : array
+        The array of the values of â at next time step, at z=j+1 and z=j+2.
+        Dimension: 2*nr.
+    j : int
+        The zeta grid coordinate: 0<=j<nz.
+    dz : float
+        Zeta step size.
+    k : int
+        The rho grid coordinate: 0<=k<nr.
+    dr : float
+        Rho step size.
+    nr : int
+        Amount of points in the rho direction.
+    dt : float
+        Tau step size.
+    k0p : float
+        k0/kp, laser wave number divided by the plasma skin depth.
+    th : float
+        Phase of the envelope at the radius, z=j.
+    th1 : float
+        Phase of the envelope at the radius, z=j+1.
+    th2 : float
+        Phase of the envelope at the radius, z=j+2.
+
+    """
+    sol = (- 2 / dt ** 2 * a[j, k]
+           - (C(-1, k, k0p, dt, dz, dr)
+              - chi() / 2
+              - 1j / dt * D(th, th1, th2, dz)) * a_old[j, k]
+           - 2 * np.exp(1j * (th - th1)) / (dz * dt)
+           * (a_new[0, k] - a_old[j + 1, k])
+           + np.exp(1j * (th - th2)) / (2 * dz * dt)
+           * (a_new[1, k] - a_old[j + 2, k]))
+    if k + 1 < nr:
+        sol -= L(1, k, dr) / 2 * a_old[j, k + 1]
+    if k > 0:
+        sol -= L(-1, k, dr) / 2 * a_old[j, k - 1]
+    return sol
+
+
+@njit()
+def TDMA(a, b, c, d):
+    """TriDiagonal Matrix Algorithm: solve a linear system Ax=b,
+    where A is a tridiagonal matrix. Source:
+    https://stackoverflow.com/questions/8733015/tridiagonal-matrix-algorithm-
+    tdma-aka-thomas-algorithm-using-python-with-nump
+
+    Parameters
+    ----------
+    a : array
+        Lower diagonal of A. Dimension: nr-1.
+    b : array
+        Main diagonal of A. Dimension: nr.
+    c : array
+        Upper diagonal of A. Dimension: nr-1.
+    d : array
+        Solution vector. Dimension: nr.
+
+    """
+    n = len(d)
+    w = np.zeros(n - 1, dtype=np.complex128)
+    g = np.zeros(n, dtype=np.complex128)
+    p = np.zeros(n, dtype=np.complex128)
+
+    w[0] = c[0] / b[0]  # MAKE SURE THAT b[0]!=0
+    g[0] = d[0] / b[0]
+
+    for i in range(1, n - 1):
+        w[i] = c[i] / (b[i] - a[i - 1] * w[i - 1])
+    for i in range(1, n):
+        g[i] = (d[i] - a[i - 1] * g[i - 1]) / (b[i] - a[i - 1] * w[i - 1])
+    p[n - 1] = g[n - 1]
+    for i in range(n - 1, 0, -1):
+        p[i - 1] = g[i - 1] - w[i - 1] * p[i]
+    return p
+
+
+@njit()
+def solve_2d(k0p, zmin, zmax, nz, rmax, nr, dt, nt, a0, aold):
```

Comment:
In the new `LaserPulse` class, this method will be called [here](https://github.com/AngelFP/Wake-T/blob/adc36a4154ee23147ccee1b6bb199cf50be25770/wake_t/physics_models/laser/laser_pulse.py#L97). There, I have renamed it to `evolve_envelope`, so this should be done here too.

Also, the inputs have changed slightly: Instead of getting `k0p`, it will receive `k0` and `kp` separately, so `kp0` has to be replaced by these two parameters, and then you can calculate `k0p = k0 / kp` inside the method. In addition, `a0` and `aold` should now be the first two inputs, followed by `chi` (which has to be added) and `k0`.